src/zocalo/service/dispatcher.py

```diff
@@ -0,0 +1,217 @@
+    # Define a base path where your recipes are located (accessed with zocalo.go -r $recipename)
+    recipe_basepath = None
+
+    def filter_load_recipes_from_files(self, message, parameters):
+        """Loads named recipes from central location and merges them into the recipe object"""
+        for recipefile in message.get("recipes", []):
+            try:
+                with open(
+                    os.path.join(self.recipe_basepath, recipefile + ".json"), "r"
+                ) as rcp:
+                    named_recipe = workflows.recipe.Recipe(recipe=rcp.read())
+            except ValueError as e:
+                raise ValueError("Error reading recipe '%s': %s", recipefile, str(e))
+            except IOError as e:
+                if e.errno == errno.ENOENT:
+                    raise ValueError(
+                        f"Message references non-existing recipe {recipefile}. Recipe path is {self.recipe_basepath}",
+                    )
+                raise
+            try:
+                named_recipe.validate()
+            except workflows.Error as e:
+                raise ValueError(
+                    "Named recipe %s failed validation. %s" % (recipefile, str(e))
+                )
+            named_recipe.apply_parameters(parameters)
+            message["recipe"] = message["recipe"].merge(named_recipe)
+        return message, parameters
+
+    def filter_apply_parameters(self, message, parameters):
+        """Fill in any placeholders in the recipe of the form {name} using the
+        parameters data structure"""
+        message["recipe"].apply_parameters(parameters)
+        return message, parameters
+
+    def hook_before_filtering(self, header, message, parameters, recipe_id):
+        """Actions to be taken before message filtering
+
+        If this function returns false then processing of this message
+        will stop
+        """
+        return_code = True
+        for entry in pkg_resources.iter_entry_points(
+            "zocalo.dispatcher.hooks_before_filtering"
+        ):
+            fn = entry.load()
+            val = fn(header, message, parameters, recipe_id, self.transport, self.log)
+            if not val:
+                return_code = False
+
+        return return_code
+
+    def hook_before_dispatch(
+        self, header, message, parameters, recipe_id, filtered_message, rw,
+    ):
+        """Actions to be taken just before dispatching message
+
+        If this function returns false then processing of this message
+        will stop
+        """
+        return_code = True
+        for entry in pkg_resources.iter_entry_points(
+            "zocalo.dispatcher.hooks_before_dispatch"
+        ):
+            fn = entry.load()
+            val = fn(
+                header,
+                message,
+                parameters,
+                recipe_id,
+                filtered_message,
+                rw,
+                self.transport,
+                self.log,
+            )
```

Comment:
Because you do stuff like this in your internal dispatcher:
https://github.com/DiamondLightSource/python-zocalo-ispyb/blob/main/zocalo_ispyb/filter.py#L418